### PR TITLE
Backports oracle/graal#2681

### PR DESCRIPTION
**Pull request by zakkak**
Backports oracle/graal#2681

### CODE REVIEW

**Comment by Karm** [comment]
Build is here: https://ci.modcluster.io/view/Mandrel/job/mandrel-linux-build/66/

And Quarkus native TS is running....will take just under 4 hours.

File: compiler/src/org.graalvm.compiler.nodes/src/org/graalvm/compiler/nodes/java/MethodCallTargetNode.java [hunk 1]
```diff
@@ -187,6 +187,11 @@ public void simplify(SimplifierTool tool) {
             // check if the type of the receiver can narrow the result
             ValueNode receiver = receiver();
 
+            ResolvedJavaType referencedReceiverType = referencedType();
```
Comment (jerboaa) [inline compiler/src/org.graalvm.compiler.nodes/src/org/graalvm/compiler/nodes/java/MethodCallTargetNode.java:190]: I wonder if we need/should drop the invokeKind().isInterface() condition on Line 185:

```
$ git diff
diff --git a/compiler/src/org.graalvm.compiler.nodes/src/org/graalvm/compiler/nodes/java/MethodCallTargetNode.java b/compiler/src/org.graalvm.compiler.nodes/src/org/graalvm/compiler/nodes/java/MethodCallTargetNode.java
index 1783483d77c..6dbf78ceccf 100644
--- a/compiler/src/org.graalvm.compiler.nodes/src/org/graalvm/compiler/nodes/java/MethodCallTargetNode.java
+++ b/compiler/src/org.graalvm.compiler.nodes/src/org/graalvm/compiler/nodes/java/MethodCallTargetNode.java
@@ -182,7 +182,7 @@ public class MethodCallTargetNode extends CallTargetNode implements IterableNode
          * Even though we are not registering an assumption (see comment below), the optimization is
          * only valid when speculative optimizations are enabled.
          */
-        if (invokeKind().isIndirect() && invokeKind().isInterface() && assumptions != null) {
+        if (invokeKind().isIndirect() && assumptions != null) {
 
             // check if the type of the receiver can narrow the result
             ValueNode receiver = receiver();
```

Comment (zakkak) [reply compiler/src/org.graalvm.compiler.nodes/src/org/graalvm/compiler/nodes/java/MethodCallTargetNode.java:190]: I think we shouldn't remove it (it's not removed upstream) https://github.com/tkrodriguez/graal/blob/topic/GR-20002/compiler/src/org.graalvm.compiler.nodes/src/org/graalvm/compiler/nodes/java/MethodCallTargetNode.java#L181

I see `invokeKind().isIndirect()` missing upstream but I think this is related to a different patch that we should backport if need be.

**Comment by zakkak** [comment]
> And Quarkus native TS is running....will take just under 4 hours.

Something looks wrong there
image

**Comment by Karm** [comment]
@zakkak  I don't think anything is wrong. It really takes single digit minutes to build Mandrel on that box. Once successfully completed, that job triggers test job and that is still running:

https://ci.modcluster.io/view/Mandrel/job/mandrel-linux-test/OS_AXIS=centos8/61/

(the Centos 7 failure is not related to this PR; I'm gonna fix the script)

File: compiler/src/org.graalvm.compiler.nodes/src/org/graalvm/compiler/nodes/java/MethodCallTargetNode.java (hunk 1 above)
Comment (jerboaa) [reply compiler/src/org.graalvm.compiler.nodes/src/org/graalvm/compiler/nodes/java/MethodCallTargetNode.java:190]: Right, my thinking was that this check be done for virtual invocations too. However, I'm not sure that's a possible path in this context. It passes the test so I'm OK for this to be merged.

**Pull request merged**